cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala

```diff
@@ -143,20 +151,76 @@ object ReportGenerationHelper {
   }
 
+  /**
+    * Generates a summary of the report with the information from the report generation request and the DataFrame
+    * containing the report details, then writes the summary to the given location.
+    *
+    * @param request the report generation request
+    * @param df the DataFrame containing report details
+    * @param reportIdDir the location to write the summary to
+    */
+  private def writeSummary(request: ReportGenerationRequest, df: DataFrame, reportIdDir: Location): Unit = {
+    val namespaces = mutable.HashSet[NamespaceAggregate]()
+    // group the report details by namespace, and then collect the count and the corresponding unique namespaces
+    df.groupBy(Constants.NAMESPACE).count.collect.foreach(r => namespaces +=
+      new NamespaceAggregate(r.getAs[String](Constants.NAMESPACE), r.getAs[Long](COUNT_COL)))
+    // group the report details by artifact information including artifact name, version and scope,
+    // and then collect the count and the corresponding unique artifact information
+    val artifacts =  mutable.HashSet[ArtifactAggregate]()
+    df.groupBy(Constants.ARTIFACT_NAME, Constants.ARTIFACT_VERSION, Constants.ARTIFACT_SCOPE).count.collect
+      .foreach(r => artifacts += new ArtifactAggregate(r.getAs[String](Constants.ARTIFACT_NAME),
+        r.getAs[String](Constants.ARTIFACT_VERSION), r.getAs[String](Constants.ARTIFACT_SCOPE),
+        r.getAs[Long](COUNT_COL)))
+    // aggregate the report details into a row with the min, max, and average of duration,
+    // and with the min and max of start
+    val aggRow = df.agg(min(df(Constants.DURATION)).as("minDuration"), max(df(Constants.DURATION)).as("maxDuration"),
+      avg(df(Constants.DURATION)).as("avgDuration"), min(df(Constants.START)).as("minStart"),
+      max(df(Constants.START)).as("maxStart")).first
+    // get the min, max, and average of duration
+    val durations = new DurationStats(aggRow.getAs[Long]("minDuration"),
+      aggRow.getAs[Long]("maxDuration"), aggRow.getAs[Double]("avgDuration"))
+    // get the min and max of start
+    val starts = new StartStats(aggRow.getAs[Long]("minStart"), aggRow.getAs[Long]("maxStart"))
+    // group the report details by the user who starts the program run, and then collect the count and
+    // the corresponding unique users
+    val owners = mutable.HashSet[UserAggregate]()
+    df.groupBy(Constants.USER).count.collect
+      .foreach(r => owners += new UserAggregate(r.getAs[String](Constants.USER), r.getAs[Long](COUNT_COL)))
+    // group the report details by the start method of the program run, and then collect the count and
+    // the corresponding unique start methods
+    val startMethods = mutable.HashSet[StartMethodAggregate]()
+    df.groupBy(Constants.START_METHOD).count.collect
+      .foreach(r => startMethods +=
+        new StartMethodAggregate(r.getAs[String](Constants.START_METHOD), r.getAs[Long](COUNT_COL)))
```

Comment:
a lot of this code is similar for grouping and counting for namespace, artifact, startMethod, owners. rather you can have some sort of list(ns, artifact, etc) and iterate through that for grouping by and for creating the object you call a method which can call appropriate create object and return.

Reply:
you can use `<? extends ProgramRunAggregate>`